### DISCUSSION TITLE
How to enable SPI on Jetson Orin Nano Devkit

### DISCUSSION THREAD

**Discussion by noob502** [OE4T/meta-tegra#1604] [Q&A] [answered]
Hello, 

I am building custom image for orin nano devkit using Yocto_Openembedded-core tools. And I am testing the SPI pins on 40 pin header. 

Following [SPI on Xavier NX](https://github.com/OE4T/meta-tegra/discussions/1223), I have added into my local.conf

```
KERNEL_MODULE_AUTOLOAD += "spidev"
KERNEL_DEVICETREE_APPLY_OVERLAYS = "tegra194-p3668-all-p3509-0000-hdr40.dtbo"
IMAGE_INSTALL:append= " kernel-module-spidev spidev-test "
```

It is not working. I did not find tegra237-p3767-all-p3768-0000-hdr40.dtb into the tools. 

I am on L4T 35.3 and meta-tegra kirkstone branch. 

Morover I also tried to modify related dtb but I did not find them in running system 

```
sudo cat /sys/kernel/debug/tegra_pinctrl_reg | grep -i spi
sudo cat /sys/kernel/debug/tegra_gpio
```

after running` ls /proc/device-tree` I got 

``` 
 spi@3210000                                                                     
 spi@3230000                                                                     
 spi@3240000                                                                     
 spi@3250000                                                                     
 spi@3300000                                                                     
 spi@c260000 
```
I found the main dtb on running system `/boot/tegra234-p3767-0003-p37668-0000-a0.dtb ` but did not find spi setting. 

What is can be the solution ?

**Comment by madisongh** [comment]
The overlay you applied is specific to the Xavier NX; it wouldn't work for Orin devices. I *think* the correct overlay to use in your case is `tegra234-p3767-0000-p3509-a02-hdr40.dtbo`, although I haven't tested it myself. The name of the file implies that it's for use with the p3509 carrier, but from what I can tell, the pin assignments on that carrier and the p3768 carrier used in the Orin Nano dev kit are the same.

**Reply by noob502** [reply]
Yes, you are right and I also figure out that the name implies for the SOM and Carrier Board though I tried if it works but did not. 

But after your reply I added `tegra234-p3767-0000-p3509-a02-hdr40.dtbo` and tested grep 'spi1' `/sys/kernel/debug/pinctrl/2430000.pinmux/pinmux-pins` and that I got 

```
# grep 'spi1' /sys/kernel/debug/pinctrl/2430000.pinmux/pinmux-pin
pin 133 (SPI1_SCK_PZ3): 2430000.pinmux (GPIO UNCLAIMED) (HOG) function spi1 group spi1_sck_pz3
pin 134 (SPI1_MISO_PZ4): 2430000.pinmux (GPIO UNCLAIMED) (HOG) function spi1 group spi1_miso_pz4
pin 135 (SPI1_MOSI_PZ5): 2430000.pinmux (GPIO UNCLAIMED) (HOG) function spi1 group spi1_mosi_pz5
pin 136 (SPI1_CS0_PZ6): 2430000.pinmux (GPIO UNCLAIMED) (HOG) function spi1 group spi1_cs0_pz6
pin 137 (SPI1_CS1_PZ7): 2430000.pinmux (GPIO UNCLAIMED) (HOG) function spi1 group spi1_cs1_pz7
```


After that I will  try to do spidev-test but still I can not find `tegra_pinctrl_reg` and `tegra_gpio`. Any idea why ?

**Reply by madisongh** [reply]
It looks like the current pinctrl driver doesn't export those extra debugfs interfaces. They probably aren't needed any more.

**Reply by noob502** [reply]
I made a test script which should look for` /dev/spidev*` but there are nothing like this.

**Reply by noob502** [reply]
I also added spidev_test which supposed to be under` tools/spidev_test` but I found it on `/usr/bin/spidev_test` and it is also looking for `/dev/spidev/` but it aborted as there is nothing like spidev* in /dev/

**Reply by madisongh** [reply, marked as answer]
Is the `spidev` kernel module loaded?

I've set this up on my Orin Nano dev kit with the following additions in my local.conf:
```
KERNEL_MODULE_AUTOLOAD:append = " spidev"
KERNEL_DEVICETREE_APPLY_OVERLAYS = "tegra234-p3767-0000-p3509-a02-hdr40.dtbo"
CORE_IMAGE_EXTRA_INSTALL = "kernel-module-spidev spidev-test"
```
These settings differ a bit from the ones in your original post.

This was for a build of the `demo-image-base` image out of the kirkstone branch of our [demo distro](https://github.com/OE4T/tegra-demo-distro).  I flashed this to my dev kit, connected pins 19 and 21 on the 40-pin header, and ran:
```
# lsmod | grep spidev
spidev                 24576  0
# spidev_test -v -D /dev/spidev0.0
spi mode: 0x0
bits per word: 8
max speed: 500000 Hz (500 kHz)
TX | FF FF FF FF FF FF 40 00 00 00 00 95 FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF F0 0D  |......@.........................|
RX | FF FF FF FF FF FF 40 00 00 00 00 95 FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF F0 0D  |......@.........................|
```
to verify that it's working.

**Reply by noob502** [reply]
I built tegra-demo-distro and also got a similar result 

```
# lsmod | grep spidev                             
spidev                 28672  0                                                 
# spidev_test -v -D /dev/spidev0.0                
spi mode: 0x0                                                                   
bits per word: 8                                                                
max speed: 500000 Hz (500 kHz)                                                  
TX | FF FF FF FF FF FF 40 00 00 00 00 95 FF FF FF FF FF FF FF FF FF FF FF FF FF|
RX | 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00|
```

But when I build with meta-tegra, meta-openembedded, and openembedded-core, I can not see the same result.

**Reply by madisongh** [reply]
Well, at least you have something working that you can compare against, to see what's set up wrong in your environment.  The demo distro includes all of those layers, and doesn't really do any magic on top of that, so it's probably something pretty basic.

**Reply by noob502** [reply]
That's a good idea. I did a quick layer comparison, but it is not clear whether those layers are dependable. But I will look at them.
